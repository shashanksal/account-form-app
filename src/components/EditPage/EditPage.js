import React from 'react';
import { useSelector } from 'react-redux';
//Material UI imports
import {
	Grid,
	Typography,
	TextField,
	TextareaAutosize,
} from '@material-ui/core';
import { createStyles, makeStyles } from '@material-ui/core/styles'; // Using Material UI styles

import { validateEmail, validateAUNumber } from '../../utils/helper';

const useStyles = makeStyles((theme) =>
	createStyles({
		form: {},
		lableFont: {
			fontWeight: 600,
			fontSize: 14,
		},
		bodyFont: {
			fontSize: 14,
		},
		rightHelperText: {
			textAlign: 'right',
		},
		textArea: {
			width: '100%',
		},
	})
);

const EditPage = (props) => {
	const [fNameError, setFNameError] = React.useState(false);
	const [lNameError, setLNameError] = React.useState(false);
	const [emailError, setEmailError] = React.useState(false);
	const [phoneError, setPhoneError] = React.useState(false);
	const [dobError, setDobError] = React.useState(false);
	const [bioError, setBioError] = React.useState(false);

	const user = useSelector((state) => state.rAccount);

	const fNameFontColor = fNameError ? '#F44335' : '#05AE95';
	const lNameFontColor = lNameError ? '#F44335' : '#05AE95';
	const emailFontColor = emailError ? '#F44335' : '#05AE95';
	const phoneFontColor = phoneError ? '#F44335' : '#05AE95';
	const dobFontColor = dobError ? '#F44335' : '#05AE95';
	const bioFontColor = bioError ? '#F44335' : '#05AE95';

	const validateFName = (event) => {
		let value = event.target.value;
		if (!value && !fNameError) {
			setFNameError(true);
		} else setFNameError(false);
	};

	const validateLName = (event) => {
		let value = event.target.value;
		if (!value && !lNameError) {
			setLNameError(true);
		} else setLNameError(false);
	};

	const validateEmailAddress = (event) => {
		let value = event.target.value;
		if ((!value && !emailError) || !validateEmail(value)) {
			setEmailError(true);
		} else setEmailError(false);
	};

	const validatePhoneNumber = (event) => {
		let value = event.target.value;
		if ((!value && !phoneError) || !validateAUNumber(value)) {
			setPhoneError(true);
		} else setPhoneError(false);
	};

	const validateDate = (event) => {
		let value = event.target.value;
		if (!value && !dobError) {
			setDobError(true);
		} else setDobError(false);
	};

	const validateBio = (event) => {
		let value = event.target.value;
		if (!value && !bioError) {
			setBioError(true);
		} else setBioError(false);
	};

	const classes = useStyles();
	return (
		<div>
			<form className={classes.form} noValidate>
				{/* First Name */}
				<TextFieldComponent
					fontColor={fNameFontColor}
					lableFont={classes.lableFont}
					title="First Name"
					error={fNameError}
					defaultValue={user.first_name}
					onChange={validateFName}
					rightHelperText={classes.rightHelperText}
					helperText="Your first name is required"
					id="fName"
				/>

				{/* Last Name */}
				<TextFieldComponent
					fontColor={lNameFontColor}
					lableFont={classes.lableFont}
					title="Last Name"
					error={lNameError}
					defaultValue={user.last_name}
					onChange={validateLName}
					rightHelperText={classes.rightHelperText}
					helperText="Your last name is required"
					id="lName"
				/>

				{/* Email */}
				<TextFieldComponent
					fontColor={emailFontColor}
					lableFont={classes.lableFont}
					title="Email"
					error={emailError}
					defaultValue={user.email}
					onChange={validateEmailAddress}
					rightHelperText={classes.rightHelperText}
					helperText="Your email is not valid email"
					id="email"
				/>

				{/* Phone Number */}
				<TextFieldComponent
					fontColor={phoneFontColor}
					lableFont={classes.lableFont}
					title="Phone"
					error={phoneError}
					defaultValue={user.phone}
					onChange={validatePhoneNumber}
					rightHelperText={classes.rightHelperText}
					helperText="Your phone number is not a valid number"
					id="phone"
				/>

				{/* Date of Birth */}
				<Grid container spacing={2}>
					<Grid item sm={12}>
						<Typography
							display="block"
							align="left"
							color="textPrimary"
							className={classes.lableFont}
							style={{ color: dobFontColor }}
						>
							Date of Birth
						</Typography>
						<TextField
							fullWidth
							id="date"
							onChange={validateDate}
							type="date"
							defaultValue={new Date(user.dob).toLocaleDateString(
								'en-CA'
							)}
							className={classes.dateField}
							InputLabelProps={{
								shrink: true,
								required: true,
							}}
						/>
					</Grid>
				</Grid>

				{/* Bio */}
				<Grid container spacing={2}>
					<Grid item sm={12}>
						<Typography
							display="block"
							align="left"
							color="textPrimary"
							className={classes.lableFont}
							style={{ color: bioFontColor }}
						>
							Date of Birth
						</Typography>
						<TextareaAutosize
							rowsMin={3}
							defaultValue={user.bio}
							id="bio"
							className={classes.textArea}
							onChange={validateBio}
							helperText={
								bioError ? (
									<Typography
										variant="caption"
										className={props.rightHelperText}
										display="block"
									>
										Please provide a bio
									</Typography>
								) : null
							}
						/>
					</Grid>
				</Grid>
			</form>
		</div>
	);
};

const TextFieldComponent = (props) => {
	return (
		<Grid container spacing={2}>
			<Grid item sm={12}>
				<Typography
					display="block"
					align="left"
					color="textPrimary"
					className={props.lableFont}
					style={{ color: props.fontColor }}
				>
					{props.title}
				</Typography>
				<TextField
					InputLabelProps={{ shrink: false }}
					error={props.error}
					fullWidth
					id={props.id}
					defaultValue={props.defaultValue}
					onChange={props.onChange}
					helperText={
						props.error ? (
							<Typography
								variant="caption"
								className={props.rightHelperText}
								display="block"
							>
								{props.helperText}
							</Typography>
						) : null
					}
				></TextField>
			</Grid>
		</Grid>
	);
};

export default EditPage;
