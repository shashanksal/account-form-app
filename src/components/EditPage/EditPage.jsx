import React from 'react';
import { useSelector , useDispatch } from 'react-redux';
//Material UI imports
import {
	Grid,
	Typography,
	TextField,
	TextareaAutosize,
} from '@material-ui/core';
import { createStyles, makeStyles } from '@material-ui/core/styles'; // Using Material UI styles

import { validateEmail, validateAUNumber } from '../../utils/helper';

import {
	storeTempLData,
	storeTempFData,
	storeTempBio,
	storeTempEmail,
	storeTempDob,
	storeTempPhone
} from '../../actions/formActions';

import{
	storeLDataErrorFlag,
	storeFDataErrorFlag,
	storeBioErrorFlag,
	storeEmailErrorFlag,
	storeDobErrorFlag,
	storePhoneErrorFlag
} from '../../actions/errorHandlerActions';

const errorColor = "#F44335";
const primaryColor = "#05AE95";

const useStyles = makeStyles((theme) =>
	createStyles({
		form: {},
		lableFont: {
			fontWeight: 600,
			fontSize: 14,
		},
		bodyFont: {
			fontSize: 14,
		},
		rightHelperText: {
			textAlign: 'right',
		},
		textArea: {
			width: '100%',
		},
	})
);

const EditPage = (props) => {
	const dispatch = useDispatch();

	// State for handling Errors
	let fNameError = useSelector((state) => state.rErrorHandler.fNameFlag);
	let lNameError = useSelector((state) => state.rErrorHandler.lNameFlag);
	let emailError = useSelector((state) => state.rErrorHandler.emailFlag);
	let phoneError = useSelector((state) => state.rErrorHandler.phoneFlag);
	let dobError = useSelector((state) => state.rErrorHandler.dobFlag);
	let bioError = useSelector((state) => state.rErrorHandler.bioFlag);

	const user = useSelector((state) => state.rAccount);

	const fNameFontColor = fNameError ?  errorColor: primaryColor;
	const lNameFontColor = lNameError ? errorColor : primaryColor;
	const emailFontColor = emailError ? errorColor : primaryColor;
	const phoneFontColor = phoneError ? errorColor : primaryColor;
	const dobFontColor = dobError ? errorColor : primaryColor;
	const bioFontColor = bioError ? errorColor : primaryColor;

	const handleFNameChange = (event) => {
		let value = event.target.value;
		if (!value && !fNameError) {
			dispatch(storeFDataErrorFlag(true))
		} else {
			dispatch(storeFDataErrorFlag(false))
			dispatch(storeTempFData(value))
		};
	};

	const handleLNameChange = (event) => {
		let value = event.target.value;
		if (!value && !lNameError) {
			dispatch(storeLDataErrorFlag(true))
		} else {
			dispatch(storeLDataErrorFlag(false))
			dispatch(storeTempLData(value))
		};
	};

	const handleEmailAddressChange = (event) => {
		let value = event.target.value;
		if ((!value && !emailError) || !validateEmail(value)) {
			dispatch(storeEmailErrorFlag(true))
		} else {
			dispatch(storeEmailErrorFlag(false))
			dispatch(storeTempEmail(value))
		};
	};

	const handlePhoneNumberChange = (event) => {
		let value = event.target.value;
		if ((!value && !phoneError) || !validateAUNumber(value)) {
			dispatch(storePhoneErrorFlag(true))
		} else {
			dispatch(storePhoneErrorFlag(false))
			dispatch(storeTempPhone(value))
		};
	};

	const handleDateChange = (event) => {
		let value = event.target.value;
		if (!value && !dobError) {
			dispatch(storeDobErrorFlag(true))
		} else {
			dispatch(storeDobErrorFlag(false))
			dispatch(storeTempDob(value))
		};
	};

	const handleBioChange = (event) => {
		let value = event.target.value;
		if (!value && !bioError) {
			dispatch(storeBioErrorFlag(true))
		} else {
			dispatch(storeBioErrorFlag(false))
			dispatch(storeTempBio(value))
		};
	};

	const classes = useStyles();
	return (
		<div>
			<form className={classes.form} noValidate id='accountForm'>
				{/* First Name */}
				<TextFieldComponent
					fontColor={fNameFontColor}
					lableFont={classes.lableFont}
					title="First Name"
					error={fNameError}
					defaultValue={user.first_name}
					onChange={handleFNameChange}
					rightHelperText={classes.rightHelperText}
					helpertext="Your first name is required"
					id="fName"
				/>

				{/* Last Name */}
				<TextFieldComponent
					fontColor={lNameFontColor}
					lableFont={classes.lableFont}
					title="Last Name"
					error={lNameError}
					defaultValue={user.last_name}
					onChange={handleLNameChange}
					rightHelperText={classes.rightHelperText}
					helpertext="Your last name is required"
					id="lName"
				/>

				{/* Email */}
				<TextFieldComponent
					fontColor={emailFontColor}
					lableFont={classes.lableFont}
					title="Email"
					error={emailError}
					defaultValue={user.email}
					onChange={handleEmailAddressChange}
					rightHelperText={classes.rightHelperText}
					helpertext="Your email is not valid email"
					id="email"
				/>

				{/* Phone Number */}
				<TextFieldComponent
					fontColor={phoneFontColor}
					lableFont={classes.lableFont}
					title="Phone"
					error={phoneError}
					defaultValue={user.phone}
					onChange={handlePhoneNumberChange}
					rightHelperText={classes.rightHelperText}
					helpertext="Your phone number is not a valid number"
					id="phone"
				/>

				{/* Date of Birth */}
				<Grid container spacing={2}>
					<Grid item sm={12}>
						<Typography
							display="block"
							align="left"
							color="textPrimary"
							className={classes.lableFont}
							style={{ color: dobFontColor }}
						>
							Date of Birth
						</Typography>
						<TextField
							fullWidth
							id="date"
							onChange={handleDateChange}
							type="date"
							defaultValue={new Date(user.dob).toLocaleDateString(
								'en-CA'
							)}
							className={classes.dateField}
							InputLabelProps={{
								shrink: true,
								required: true,
							}}
						/>
					</Grid>
				</Grid>

				{/* Bio */}
				<Grid container spacing={2}>
					<Grid item sm={12}>
						<Typography
							display="block"
							align="left"
							color="textPrimary"
							className={classes.lableFont}
							style={{ color: bioFontColor }}
						>
							Date of Birth
						</Typography>
						<TextareaAutosize
							rowsMin={3}
							defaultValue={user.bio}
							id="bio"
							className={classes.textArea}
							onChange={handleBioChange}
							helpertext={
								bioError ? (
									<Typography
										variant="caption"
										className={props.rightHelperText}
										display="block"
									>
										Please provide a bio
									</Typography>
								) : null
							}
						/>
					</Grid>
				</Grid>
			</form>
		</div>
	);
};

const TextFieldComponent = (props) => {
	return (
		<Grid container spacing={2}>
			<Grid item sm={12}>
				<Typography
					display="block"
					align="left"
					color="textPrimary"
					className={props.lableFont}
					style={{ color: props.fontColor }}
				>
					{props.title}
				</Typography>
				<TextField
					InputLabelProps={{ shrink: false }}
					error={props.error}
					fullWidth
					id={props.id}
					defaultValue={props.defaultValue}
					onChange={props.onChange}
					helperText={
						props.error ? (
							<Typography
								variant="caption"
								className={props.rightHelperText}
								display="block"
							>
								{props.helpertext}
							</Typography>
						) : null
					}
				></TextField>
			</Grid>
		</Grid>
	);
};

export default EditPage;
