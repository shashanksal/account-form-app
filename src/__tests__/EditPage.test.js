import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";

import EditPage from "../components/EditPage";

const middlewares = [];
const mockStore = configureMockStore(middlewares);

const initData = {
	rAccount: {
		first_name: "Bruce",
		last_name: "Banner",
		email: "bruce.banner@marvel.com",
		phone: "0400000000",
		dob: new Date("31 May 1970"),
		bio:
			'Dr. Bruce Banner is a character in the Marvel Cinematic Universe (MCU) film franchise initially portrayed by Edward Norton and subsequently by Mark Ruffalo—based on the Marvel Comics character of the same name—known commonly by his alter ego, the Hulk. In the films, Banner is a renowned physicist who subjected himself to a gamma radiation experiment designed to replicate a World War II-era "super soldier" program.'
	},
	rForm: false
};

const mStore = mockStore(initData);

describe("test Edit page", () => {
	it("Edit page matches snapshot", () => {
		const { asFragment } = render(
			<Provider store={mStore}>
				<EditPage />
			</Provider>
		);
		expect(asFragment()).toMatchSnapshot();
	});
});
