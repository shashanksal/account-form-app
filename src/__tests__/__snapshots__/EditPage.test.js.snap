// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Edit page Edit page matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form
      class="makeStyles-form-1"
      novalidate=""
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12"
        >
          <p
            class="MuiTypography-root makeStyles-lableFont-2 MuiTypography-body1 MuiTypography-colorTextPrimary MuiTypography-alignLeft MuiTypography-displayBlock"
            style="color: rgb(5, 174, 149);"
          >
            First Name
          </p>
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                id="fName"
                type="text"
                value="Bruce"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12"
        >
          <p
            class="MuiTypography-root makeStyles-lableFont-2 MuiTypography-body1 MuiTypography-colorTextPrimary MuiTypography-alignLeft MuiTypography-displayBlock"
            style="color: rgb(5, 174, 149);"
          >
            Last Name
          </p>
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                id="lName"
                type="text"
                value="Banner"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12"
        >
          <p
            class="MuiTypography-root makeStyles-lableFont-2 MuiTypography-body1 MuiTypography-colorTextPrimary MuiTypography-alignLeft MuiTypography-displayBlock"
            style="color: rgb(5, 174, 149);"
          >
            Email
          </p>
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                id="email"
                type="text"
                value="bruce.banner@marvel.com"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12"
        >
          <p
            class="MuiTypography-root makeStyles-lableFont-2 MuiTypography-body1 MuiTypography-colorTextPrimary MuiTypography-alignLeft MuiTypography-displayBlock"
            style="color: rgb(5, 174, 149);"
          >
            Phone
          </p>
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                id="phone"
                type="text"
                value="0400000000"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12"
        >
          <p
            class="MuiTypography-root makeStyles-lableFont-2 MuiTypography-body1 MuiTypography-colorTextPrimary MuiTypography-alignLeft MuiTypography-displayBlock"
            style="color: rgb(5, 174, 149);"
          >
            Date of Birth
          </p>
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                id="date"
                type="date"
                value="5/31/1970"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12"
        >
          <p
            class="MuiTypography-root makeStyles-lableFont-2 MuiTypography-body1 MuiTypography-colorTextPrimary MuiTypography-alignLeft MuiTypography-displayBlock"
            style="color: rgb(5, 174, 149);"
          >
            Date of Birth
          </p>
          <textarea
            class="makeStyles-textArea-5"
            id="bio"
            rows="3"
            style="height: -4px; overflow: hidden;"
          >
            Dr. Bruce Banner is a character in the Marvel Cinematic Universe (MCU) film franchise initially portrayed by Edward Norton and subsequently by Mark Ruffalo—based on the Marvel Comics character of the same name—known commonly by his alter ego, the Hulk. In the films, Banner is a renowned physicist who subjected himself to a gamma radiation experiment designed to replicate a World War II-era "super soldier" program.
          </textarea>
          <textarea
            aria-hidden="true"
            class="makeStyles-textArea-5"
            readonly=""
            style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;"
            tabindex="-1"
          />
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
